pkg/controllers/cloud/virtualmachine_controller.go

```diff
@@ -83,45 +64,111 @@ func (r *VirtualMachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		Ch:     make(chan watch.Event, converterChannelBuffer),
 		Scheme: r.Scheme,
 	}
-	if err := ctrl.NewControllerManagedBy(mgr).For(&cloudv1alpha1.VirtualMachine{}).Complete(r); err != nil {
-		return err
-	}
-	go r.converter.Start()
 	return mgr.Add(r)
 }
 
 // Start performs the initialization of the controller.
-// A controller is said to be initialized only when the dependent controllers
-// are synced, and it keeps a count of pending CRs to be reconciled.
+// VM controller is said to be initialized and synced when all the VM and EE
+// CRs are reconciled.
 func (r *VirtualMachineReconciler) Start(context.Context) error {
-	if err := GetControllerSyncStatusInstance().waitForControllersToSync([]controllerType{ControllerTypeEE}, syncTimeout); err != nil {
-		r.Log.Error(err, "dependent controller sync failed", "controller", ControllerTypeEE.String())
+	if err := GetControllerSyncStatusInstance().waitForControllersToSync(
+		[]controllerType{ControllerTypeEE, ControllerTypeCES, ControllerTypeCPA}, syncTimeout); err != nil {
 		return err
 	}
-	vmList := &cloudv1alpha1.VirtualMachineList{}
-	if err := r.Client.List(context.TODO(), vmList, &client.ListOptions{}); err != nil {
+	r.Log.Info("Init done", "controller", ControllerTypeVM.String())
+
+	/* TODO: Handle restart case. Fetch all EE and all VMs (in a map) and walk thru EE and check if it exists
+	 * in the map. If not, issues a delete for that EE.
+	 */
+	if err := r.syncExternalEntities(); err != nil {
 		return err
 	}
 
-	r.pendingSyncCount = len(vmList.Items)
-	if r.pendingSyncCount == 0 {
-		GetControllerSyncStatusInstance().SetControllerSyncStatus(ControllerTypeVM)
+	// Start the converter module.
+	go r.converter.Start()
+	// Set sync status as complete.
+	GetControllerSyncStatusInstance().SetControllerSyncStatus(ControllerTypeVM)
+	// Setup watcher after VM cache is populated and EE CRs are reconciled.
+	if err := r.resetWatcher(); err != nil {
+		return err
+	}
+	// Blocking thread to wait for any VM event.
+	if err := r.processEvent(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getAllVMs returns a map which contains all the VMs.
+func (r *VirtualMachineReconciler) getAllVMs() map[types.NamespacedName]struct{} {
+	vmObjList := r.Inventory.GetAllVms()
+	tempVmMap := map[types.NamespacedName]struct{}{}
+	for _, vmObj := range vmObjList {
+		vm := vmObj.(*runtimev1alpha1.VirtualMachine)
+		vmNn := types.NamespacedName{
+			Name:      vm.Name,
+			Namespace: vm.Namespace,
+		}
+		tempVmMap[vmNn] = struct{}{}
+	}
+	return tempVmMap
+}
+
+// syncExternalEntities validates that each EE has corresponding VM. If it does not exist then
+// the EE will be deleted.
+func (r *VirtualMachineReconciler) syncExternalEntities() error {
+	vmMap := r.getAllVMs()
+	eeList := &antreav1alpha2.ExternalEntityList{}
+	if err := r.Client.List(context.TODO(), eeList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	for _, ee := range eeList.Items {
```

Comment:
check if nephe label exists. if not, consider is non nephe managed EE

Reply:
Done